File: java/unique_binary_search_trees_ii.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.Stack;
import java.util.PriorityQueue;
import java.lang.Math;

public class HelloWorld {    
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }
    public List<TreeNode> generateTrees(int l, int r) {
        List<TreeNode> res=new ArrayList<TreeNode>();
        if (l>r) {
            res.add(null);
            return res;
        }
        for (int i=l; i<=r; i++) {            
            List<TreeNode> left=generateTrees(l, i-1);
            List<TreeNode> right=generateTrees(i+1, r);
            for (int j=0; j<left.size(); j++) {
                for (int k=0; k<right.size(); k++) {
                    TreeNode node=new TreeNode(i);
                    node.left=left.get(j);
                    node.right=right.get(k);
                    res.add(node);
                }
            }                       
        }
        
        return res;
    }
    
    public List<TreeNode> generateTrees(int n) {
        if (n==0) return new ArrayList<TreeNode> ();
        return generateTrees(1, n);
    }
         
    public static void main(String[] args) {
    }
}

File: cpp/unique_binary_search_trees_II.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <string>
#include <vector>
#include <queue>
#include <stack>
#include <unordered_set>
#include <map>
#include <limits.h>

using namespace std;

/**
 * Definition for binary tree
 */
struct TreeNode {
     int val;
     TreeNode *left;
     TreeNode *right;
     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

/**
 * Definition for singly-linked list.
 */
struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(NULL) {}
 };

/**
 * Definition for undirected graph.
 * */
struct UndirectedGraphNode {
    int label;
    vector<UndirectedGraphNode *> neighbors;
    UndirectedGraphNode(int x) : label(x) {};
};

/**
 * Definition for binary tree with next pointer.
 */
struct TreeLinkNode {
  int val;
  TreeLinkNode *left, *right, *next;
  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
};

class Solution {
public:
    vector<TreeNode*> generateTrees(int l, int r) {
        vector<TreeNode*> res;
        if (l>=r) {
            res.push_back(l==r?new TreeNode(l):NULL);
            return res;
        }
        for (int i=l; i<=r; i++) {
            vector<TreeNode*> left=generateTrees(l, i-1);
            vector<TreeNode*> right=generateTrees(i+1, r);
            for (int j=0; j<left.size(); j++) {
                for (int k=0; k<right.size(); k++) {
                    TreeNode *n=new TreeNode(i);
                    n->left=left[j];
                    n->right=right[k];
                    res.push_back(n);
                }
            }
        }
        
        return res;
    }
    vector<TreeNode*> generateTrees(int n) {
        if (n==0) return vector<TreeNode*>();
        return generateTrees(1, n);
    }
};

int main()
{
	return 0;
}

